tensorbay/client/requests.py

```diff
@@ -303,46 +306,118 @@ def multithread_upload(
             future.result()
 
 
+PagingGenerator = Callable[[int, int], Generator[_T, None, int]]
+
+
+class LazyItem(Generic[_T]):  # pylint: disable=too-few-public-methods
+    """In paging lazy evaluation system, a LazyItem instance represents an element in a pagination.
+
+    If user wants to access the elememt, LazyItem will trigger the paging request to pull a page of
+    elements and return the required element. All the pulled elements will be stored in different
+    LazyItem instances and will not be requested again.
+
+    Arguments:
+        page: The page the item belongs to.
+
+    """
+
+    __slots__ = ("data", "_page")
+
+    data: _T
+
+    def __init__(self, page: "LazyPage[_T]"):
+        self._page = page
+
+    def get(self) -> _T:
+        """Access the actual element represented by LazyItem.
+
+        If the element is already pulled from web, it will be return directly, otherwise this
+        function will request for a page of elements to get the required elememt.
+
+        Returns:
+            The actual element this LazyItem instance represents.
+
+        """
+        if not hasattr(self, "data"):
+            self._page.pull()
+
+        return self.data
+
+
+class LazyPage(Generic[_T]):  # pylint: disable=too-few-public-methods
+    """In paging lazy evaluation system, a LazyPage instance represents a page with elements.
+
+    LazyPage is used for sending paging request to pull a page of elements and storing them in
+    different :class:`LazyItem` instances.
+
+    Arguments:
+        offset: The offset of the page.
+        limit: The limit of the page.
+        func: A paging generator function, which takes offset<int> and limit<int> as inputs and
+            returns a generator. The returned generator should yield the element user needs, and
+            return the total count of the elements in the paging request.
+
+    """
+
+    __slots__ = ("_offset", "_limit", "_func", "_lock", "_total_count", "items")
+
+    def __init__(self, offset: int, limit: int, func: PagingGenerator[_T]) -> None:
+        self._offset = offset
+        self._limit = limit
+        self._func = func
+        self._lock = Lock()
+        self._total_count: Optional[int] = None
+
+        self.items: Tuple[LazyItem[_T], ...] = tuple(LazyItem(self) for _ in range(limit))
+
+    def pull(self) -> int:
+        """Send paging request to pull a page of elements and store them in :class:`LazyItem`.
+
+        Returns:
+            The totalCount of all elements.
+
+        """
+        with self._lock:
+            if self._total_count is None:
+                iterator = zip(self._func(self._offset, self._limit), self.items)
+                try:
+                    while True:
```

Comment:
```
for data, item in iterator:
    item.data = data
self._total_count = ...
```

 maybe be better cuz its no need to catch "StopIteration"

Reply:
Because the return of the Generator is needed. Catch `StopIteration` is for getting the return value.